x-pack/plugins/security_solution/public/common/components/filter_group/utils.ts

```diff
@@ -30,6 +33,115 @@ export const getFilterItemObjListFromControlInput = (controlInput: ControlGroupI
+   *
+   */
+  controlsWithPriority: FilterItemObj[][];
+  defaultControlsObj: Record<string, FilterItemObj>;
+}
+
+/*
+ * mergeControls merges controls based on priority with the default controls
+ *
+ * @return undefined if all provided controls are empty
+ * */
+export const mergeControls = ({
+  controlsWithPriority,
+  defaultControlsObj,
+}: MergableControlsArgs) => {
+  const highestPriorityControlSet = controlsWithPriority.find((control) => !isEmpty(control));
+
+  return highestPriorityControlSet?.map((singleControl) => {
+    if (singleControl.fieldName in defaultControlsObj) {
+      return {
+        ...defaultControlsObj[singleControl.fieldName],
+        ...singleControl,
+      };
+    }
+    return singleControl;
+  });
+};
+
+interface ReorderControlsArgs {
+  /*
+   * Ordered Controls
+   *
+   * */
+  controls: FilterItemObj[];
+  /*
+   * default controls in order
+   * */
+  defaultControls: FilterItemObj[];
+}
+
+/**
+ * reorderControlsWithPersistentControls reorders the controls such that controls which
+ * are persistent in default controls should be upserted in given order
+ *
+ * */
+export const reorderControlsWithDefaultControls = (args: ReorderControlsArgs) => {
+  const { controls, defaultControls } = args;
+  const controlsObject = controls.reduce((prev, current) => {
+    prev[current.fieldName] = current;
+    return prev;
+  }, {} as Record<string, FilterItemObj>);
+
+  const defaultControlsObj = defaultControls.reduce((prev, current) => {
+    prev[current.fieldName] = current;
+    return prev;
+  }, {} as Record<string, FilterItemObj>);
+
+  const resultDefaultControls: FilterItemObj[] = defaultControls
+    .filter((defaultControl) => defaultControl.persist)
+    .map((defaultControl) => {
+      return {
+        ...defaultControl,
+        ...(controlsObject[defaultControl.fieldName] ?? {}),
+      };
+    });
+
+  const resultNonPersitantControls = controls
+    .filter(
+      // filter out persisting controls since we have already taken
+      // in account above
+      (control) => !defaultControlsObj[control.fieldName]?.persist
+    )
+    .map((control) => ({
+      // insert some default properties from default controls
+      // irrespective of whether they are persistent or not.
+      ...(defaultControlsObj[control.fieldName] ?? {}),
+      ...control,
+    }));
+
+  return [...resultDefaultControls, ...resultNonPersitantControls];
+};
+
+/*
+ * getFilterControlsComparator provides a comparator that can be used with `isEqualWith` to compare
+ * 2 instances of FilterItemObj
+ *
+ * */
+export const getFilterControlsComparator =
+  (...fieldsToCompare: Array<keyof FilterItemObj>) =>
+  (value1: FilterItemObj[], value2: FilterItemObj[]) => {
+    if (value1.length !== value2.length) return false;
+    const valmod1 = value1.map((v) => {
+      return pick(v, fieldsToCompare);
+    });
+
+    const valmod2 = value2.map((v) => {
+      return pick(v, fieldsToCompare);
+    });
```

Comment:
is there an option for better naming here? :D does array order matter here?

Reply:
@lgestc , these are abstract array without much knowledge of domain they are working in.. But I understand what you are saying.. let me think more on name.

Reply:
@lgestc , this is done.